syntax = "proto3";

import "vgeodb_common.proto";

package VgeoDB;

option java_package = "vgeodb.localfeature";
option java_outer_classname = "lfdb";


message SuperPointDesc_t {
	enum eDescType {
		SPP_512_v1 = 0;
	}
	repeated float arfDescriptor = 2 [packed=true];
	eDescType eType = 3;
}

message KP2DDesc_t {
	enum eDescType {
		KP2D_256_v1 = 0;
	}
	repeated float arfDescriptor = 2 [packed=true];
	eDescType eType = 3;
}

message LocalFeatureDesc {
	oneof uLocalFeature_t {
		SuperPointDesc_t stSPPDesc = 3;
		KP2DDesc_t stKP2DDesc = 4;
	}
}

message LocalFeatMeta_t {
	uint64 un64UniqueID			= 1;
	uint64 un64KeyPointCount	= 2;
	uint64 un64GroupID			= 3;

	VgeoDB.common.XYZ64_t stRefLocation = 4;
}

message KeyPoint_t {
	uint64 un64Index = 1;
	uint64 un64Timestamp = 2;

	VgeoDB.common.XYZ32_t stLocation = 3;

	LocalFeatureDesc uDescriptor = 4;

	repeated uint64 arun64KeyframeUniqueID = 5 [packed=true];
	repeated uint64 arun64KeyframeIndex = 6 [packed=true];
	repeated VgeoDB.common.UVPoint_t arstKeyPointImage = 7;

	float fSimilarity = 11;
}

message LocalFeature {
	LocalFeatMeta_t stMeta = 1;
	repeated KeyPoint_t stKeypoint = 2;
}

message LocalFeatureSet {
	repeated LocalFeature arstLocalFeature = 1;
}
